src/filters.rs

```diff
@@ -506,65 +508,103 @@ pub struct BboxFilter<'a> {
-        let u64_slice =
-            unsafe { std::slice::from_raw_parts(slice.as_ptr() as *const u64, slice.len() / 8) };
-        u64_slice.to_vec()
+        slice
+            .chunks_exact(8)
+            .map(|chunk| {
+                u64::from_le_bytes(
+                    chunk
+                        .try_into()
+                        .expect("array path entries are multiples of 8 bytes"),
+                )
+            })
+            .collect()
     }
 }
 
 impl<'a> QueryRuntimeFilter for BboxFilter<'a> {
     fn first_result(&mut self, start: &DocResult) -> Option<DocResult> {
-        let query = self
-            .kb
-            .rtree_query_key(start.seq, std::u64::MAX, &self.bbox);
+        // For array queries, we need to consider the context
+        let min_seq = if self.kb.arraypath.is_empty() {
+            // Not an array query, use normal start seq
+            start.seq
+        } else if start.seq > 0 && start.arraypath.len() == self.kb.arraypath.len() {
+            // Array query for a specific document - search within that document
+            // by using seq-1 to ensure we get all array elements
+            start.seq.saturating_sub(1)
+        } else {
+            start.seq
+        };
+        
+        let query = self.kb.rtree_query_key(min_seq, std::u64::MAX, &self.bbox);
         self.iter = Some(self.snapshot.new_rtree_iterator(&query));
+        // Reset last_returned to handle the start properly
+        self.last_returned = None;
         self.next_result()
     }
 
     fn next_result(&mut self) -> Option<DocResult> {
         let iter = self.iter.as_mut().unwrap();
-        if let Some((key, value)) = iter.next() {
-            let mut vec = Vec::with_capacity(key.len());
-            vec.extend_from_slice(&key);
-            let mut read = Cursor::new(vec);
-            let key_len = read.read_unsigned_varint_32().unwrap();
-            let offset = read.position() as usize;
-
-            let iid = unsafe {
-                let array = *(key[offset + key_len as usize..].as_ptr() as *const [_; 8]);
-                mem::transmute::<[u8; 8], u64>(array)
-            };
+        
+        loop {
+            if let Some((key, value)) = iter.next() {
+                let mut vec = Vec::with_capacity(key.len());
+                vec.extend_from_slice(&key);
+                let mut read = Cursor::new(vec);
+                let key_len = read.read_unsigned_varint_32().unwrap();
+                let offset = read.position() as usize;
+
+                let iid = u64::from_le_bytes(
+                    key[offset + key_len as usize..offset + key_len as usize + 8]
+                        .try_into()
+                        .expect("iid bytes"),
+                );
+
+                // Get arraypath for this result
+                let arraypath = BboxFilter::from_u8_slice(&value);
+                
+                // Skip if we've already returned this exact result
```

Comment:
It's been a very long time since I've been deep in the Noise code base. So it's a bit as if I read it with a fresh eye. Could you please add a comment when/why this case is happening?